### PR TITLE
NIFI-6685: Add align and distribute UI actions

### PR DESCRIPTION
**Pull request by gschaer**
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

Enables align and distribute actions on the core UI. When at least two components are selected, the user can now align them vertically or horizontally. When at least three components are selected, the user can now distribute them evenly.

Carbon Icons library was added to illustrate the actions in the UI. (Apache License 2.0)
This fixes bug NIFI-6685

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [X] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [X] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [X] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [X] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [X] Have you verified that the full build is successful on JDK 8?
- [X] Have you verified that the full build is successful on JDK 11?
- [X] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by scottyaslan** [comment]
reviewing...

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/pom.xml [hunk 1]
```diff
@@ -246,6 +246,11 @@
                                         <include>moment/README.md</include>
                                         <include>moment/LICENSE</include>
 
+                                        <!-- carbon icons -->
+                                        <include>@carbon/icons/svg/**/*</include>
```
Comment (scottyaslan) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/pom.xml:250]: It would be great if we did not need to add yet another dependency here. Maybe we can use one of the available icons from font awesome or the custom flowfont created for this project?

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-context-menu.js [hunk 2]
```diff
@@ -844,9 +853,21 @@
         {id: 'show-source-menu-item', condition: isConnection, menuItem: {clazz: 'fa fa-long-arrow-left', text: 'Go to source', action: 'showSource'}},
         {id: 'show-destination-menu-item', condition: isConnection, menuItem: {clazz: 'fa fa-long-arrow-right', text: 'Go to destination', action: 'showDestination'}},
         {separator: true},
-        {id: 'align-menu-item', groupMenuItem: {clazz: 'fa', text: 'Align'}, menuItems: [
-            {id: 'align-horizontal-menu-item', condition: canAlign, menuItem: { clazz: 'fa fa-align-center fa-rotate-90', text: 'Horizontally', action: 'alignHorizontal'}},
-            {id: 'align-vertical-menu-item', condition: canAlign, menuItem: {clazz: 'fa fa-align-center', text: 'Vertically', action: 'alignVertical'}}
+        {id: 'align-menu-item', groupMenuItem: {clazz: 'carbon carbon-align--horizontal-left', text: 'Align'}, menuItems: [
```
Comment (scottyaslan) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-context-menu.js:856]: All other icons in the context menu have a color of #004849. Please update these icons to also have this color applied.

**Review by scottyaslan — CHANGES_REQUESTED** [review]

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/pom.xml (hunk 1 above)
Comment (gschaer) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/pom.xml]: Good point. #af43e5c integrates carbon icon directly into the flowfont.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-context-menu.js (hunk 2 above)
Comment (gschaer) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-context-menu.js]: By moving the carbon icons to the flowfont with commit af43e5c solves this.

**Comment by thenatog** [comment]
I believe the Hive errors in the github actions can be resolved by rebasing on main.

**Comment by gschaer** [comment]
As @thenatog mentioned, the Hive errors in the github actions have been fixed in commit 4888eaf62d2026067c40caa91e7bf6d4f2917c9f

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/resources/META-INF/NOTICE [hunk 3]
```diff
@@ -1,8 +1,17 @@
...
 The Apache Software Foundation (http://www.apache.org/).
 
+******************
+Apache Software License v2
+******************
+The following binary components are provided under the Apache Software License v2
+  (ASLv2) Carbon Design System Icons
+    The following NOTICE information applies:
+      Copyright 2015 IBM Corp.
+
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/resources/META-INF/NOTICE:14]: Is this addition still necessary with the change to flowfont?

Comment (gschaer) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/resources/META-INF/NOTICE:14]: Good question. I took the icons I needed from the Carbon Design System Icons and I integrated them into Flowfont using Fontello. Fontello was used 5 years ago to generate Flowfont: https://fontello.com/

So technically, I used some components from Carbon Design System Icons. As I understand (but I'm not a lawyer), I should add the license of the Carbon Design System Icons.

If you look at the main license file, we reference the Fontello license.  https://github.com/apache/nifi/blob/7fe8858d22f580ad9dcd96ebafb59dbe5e73f5cb/nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/resources/META-INF/LICENSE#L681-L701

As we reference the tool used to generate the font, I think we should reference the license of the icons used in the app. 

Or we need to think about removing the reference to the Fontello license.

What do you think?

Comment (exceptionfactory) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/resources/META-INF/NOTICE:14]: Thanks for the thorough response and description.  I am not an expert when it comes to licensing, but since you clarified that the Flowfont file includes icons from the Carbon Design System, I think the updated notice you included makes sense.
